src/pmdas/bpf/modules/bashreadline.c

```diff
@@ -0,0 +1,394 @@
+/*
+ *
+ * Copyright (c) 2022 Sohaib Mohamed <sohaib.amhmd@gmail.com>
+ *
+ * Based on the bashreadline(8):
+ * https://github.com/iovisor/bcc/blob/master/libbpf-tools/bashreadline.c
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; either version 2 of the License, or (at your
+ * option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * for more details.
+ */
+
+#include "module.h"
+
+#include <bpf/bpf.h>
+#include <bpf/libbpf.h>
+#include <pcp/pmda.h>
+#include <pcp/pmwebapi.h>
+#include <sys/queue.h>
+
+#include "bashreadline.h"
+#include "bashreadline.skel.h"
+#include "btf_helpers.h"
+#include "trace_helpers.h"
+#include "uprobe_helpers.h"
+
+#define PERF_BUFFER_PAGES 16
+#define PERF_POLL_TIMEOUT_MS 100
+#define INDOM_COUNT 2
+
+static struct env {
+    int process_count;
+} env = {
+    .process_count = 20,
+};
+
+static char *libreadline_path = NULL;
+
+static pmdaInstid *bashreadline_instances, *bashreadline_lost;
+static struct bashreadline_bpf *obj;
+static struct perf_buffer *pb = NULL;
+static char *readline_so_path;
+static int lost_events;
+static int queuelength;
+
+/* cache array */
+struct tailq_entry {
+    struct str_t str_t;
+    TAILQ_ENTRY(tailq_entry) entries;
+};
+
+TAILQ_HEAD(tailhead, tailq_entry) head;
+
+static struct tailq_entry* allocElm(void)
+{
+    return malloc(sizeof(struct tailq_entry));
+}
+
+static void push(struct tailq_entry *elm)
+{
+    if (queuelength > env.process_count)
+    {
+        struct tailq_entry *l;
+        l = head.tqh_first;
+        TAILQ_REMOVE(&head, l, entries);
+        free(l);
+        queuelength--;
+    }
+    TAILQ_INSERT_TAIL(&head, elm, entries);
+    queuelength++;
+}
+
+static bool get_item(unsigned int offset, struct tailq_entry** val)
+{
+    struct tailq_entry *i;
+    unsigned int iter = 0;
+
+    TAILQ_FOREACH_REVERSE(i, &head, tailhead, entries) {
+        if (offset == iter) {
+            *val = i;
+            return true;
+        }
+        iter++;
+    }
+    return false;
+}
+
+static unsigned int indom_id_mapping[INDOM_COUNT];
+
+#define METRIC_COUNT 3
+enum metric_name { COMM, PID, LOST };
+enum metric_indom { BASHREADLINE_INDOM, LOST_EVENTS };
+
+char* metric_names[METRIC_COUNT] = {
```

Comment:
const char

Reply:
@jasonk000 So, I think I need change `char* (*metric_name_fn_t)` to `const char* (*metric_name_fn_t)`?

Reply:
I think this one was for @a3f ?

Reply:
Oh, sorry ^^

Reply:
Yes.

`metric_text_oneline` and `metric_text_long` could also be constants, however they get passed down to `dp->version.four.text` (in `bpf.c`) which expects a `char**`